packages/assets-controllers/src/DeFiPositionsController/DeFiPositionsController.ts

```diff
@@ -0,0 +1,240 @@
+  typeof controllerName,
+  DeFiPositionsControllerState,
+  DeFiPositionsControllerMessenger
+> {
+  readonly #fetchPositions: (
+    accountAddress: string,
+  ) => Promise<DefiPositionResponse[]>;
+
+  /**
+   * Tokens controller options
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - The controller messenger.
+   * @param options.apiUrl - Override for the API URL to use for fetching DeFi positions.
+   * @param options.interval - Override for the interval to use for polling DeFi positions. (default: 10 minutes)
+   * @param options.isEnabled - Whether the controller is enabled. (default: true)
+   */
+  constructor({
+    messenger,
+    apiUrl,
+    interval = TEN_MINUTES_IN_MS,
+    isEnabled = () => true,
+  }: {
+    messenger: DeFiPositionsControllerMessenger;
+    apiUrl?: string;
+    interval?: number;
+    isEnabled?: () => boolean;
+  }) {
+    super({
+      name: controllerName,
+      metadata: controllerMetadata,
+      messenger,
+      state: getDefaultDefiPositionsControllerState(),
+    });
+
+    this.setIntervalLength(interval);
+
+    this.#fetchPositions = buildPositionFetcher(apiUrl);
+
+    this.messagingSystem.subscribe('KeyringController:unlock', async () => {
+      if (!isEnabled()) {
+        return;
+      }
+
+      this.startPolling(null);
+    });
+
+    this.messagingSystem.subscribe('KeyringController:lock', () => {
+      if (!isEnabled()) {
+        return;
+      }
+
+      this.stopAllPolling();
+    });
+
+    this.messagingSystem.subscribe(
+      'TransactionController:transactionConfirmed',
+      async (transactionMeta) => {
+        if (!isEnabled()) {
+          return;
+        }
+
+        const accountAddress = transactionMeta.txParams.from;
+        await this.#updateAccountPositions(accountAddress);
+      },
+    );
+  }
+
+  async _executePoll(): Promise<void> {
+    const accounts = this.messagingSystem.call(
+      'AccountsController:listAccounts',
+    );
+
+    const results = await reduceInBatchesSerially({
+      initialResult: [] as {
+        accountAddress: string;
+        positions: GroupedPositionsPerChain | null;
+      }[],
+      values: accounts,
+      batchSize: FETCH_POSITIONS_BATCH_SIZE,
+      eachBatch: async (workingResult, batch) => {
+        const batchResults = (
+          await Promise.all(
+            batch.map(async ({ address: accountAddress, type }) => {
+              if (type.startsWith('eip155:')) {
+                try {
+                  const positions =
+                    await this.#fetchAccountPositions(accountAddress);
+
+                  return {
+                    accountAddress,
+                    positions,
+                  };
+                  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+                } catch (error) {
+                  return {
+                    accountAddress,
+                    positions: null,
```

Comment:
So... I was wondering what to do if one of the calls errors and I think this is a good compromise for now.

If we have a fetch from the API failing, instead of creating a new error field for which we have no use, we can set the positions to `null` so that we can display a generic message in the client.